torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -542,7 +546,22 @@ void ProcessGroupAgent::listenLoop() {
       recvWork_ = work;
     }
 
-    if (!rpcRunning_.load() || !work->wait() /* aborted */) {
+    if (!rpcRunning_.load()) {
+      return;
+    }
+    bool aborted;
+    try {
+      aborted = !work->wait();
+    } catch (const gloo::IoException& e) {
```

Comment:
As I mentioned in https://github.com/pytorch/pytorch/pull/32602 can we use the `listenLoop` and `listenLoopInternal` mechanism to catch all exceptions? For example, `recvAnysource` on line 543 can throw exceptions as well, but we don't catch them in this PR.

As you mentioned in https://github.com/pytorch/pytorch/pull/32602, setting rpcRunning_ to false is not a good idea, but let's set some error flag in ProcessGroupAgent and call ProcessGroupAgent::shutdown to shutdown the RPC framework. Then in ProcessGroupAgent.send() we can throw an appropriate exception if listenLoop crashed due to an exception.

Reply:
Not sure if it will work to call `::shutdown()` from the listening thread. `::shutdown()` joins the listening thread itself, which could lead to issues since we indirectly have a thread call join on itself.